can't compile on Linux

after doing `cmake .` i execute the created `MakeFile` by doing `make` but i get these errors :

[ i'm on XUbuntu 22.04 ]

```
$ make
[  7%] Building CXX object CMakeFiles/haVoc.dir/bitboards.cpp.o
[ 14%] Building CXX object CMakeFiles/haVoc.dir/evaluate.cpp.o
In file included from /home/roelof/Compiled/chess-develop/evaluate.h:11,
                 from /home/roelof/Compiled/chess-develop/evaluate.cpp:5:
/home/roelof/Compiled/chess-develop/parameter.h:26:23: error: expected ‘)’ before ‘&&’ token
   26 |         parameter<T>(T&& in, std::string& s) : tag(s)
      |                     ~ ^~
      |                       )
/home/roelof/Compiled/chess-develop/parameter.h:31:23: error: expected ‘)’ before ‘&’ token
   31 |         parameter<T>(T& in, std::string& s) : tag(s) { set(in); }
      |                     ~ ^
      |                       )
/home/roelof/Compiled/chess-develop/parameter.h:32:23: error: expected ‘)’ before ‘&’ token
   32 |         parameter<T>(T& in) : tag("") { set(in); }
      |                     ~ ^
      |                       )
/home/roelof/Compiled/chess-develop/parameter.h:33:22: error: expected unqualified-id before ‘const’
   33 |         parameter<T>(const parameter<T>& o) { tag = o.tag;  set(*o.value); }
      |                      ^~~~~
/home/roelof/Compiled/chess-develop/parameter.h:33:22: error: expected ‘)’ before ‘const’
   33 |         parameter<T>(const parameter<T>& o) { tag = o.tag;  set(*o.value); }
      |                     ~^~~~~
      |                      )
/home/roelof/Compiled/chess-develop/parameter.h:34:17: error: template-id not allowed for destructor
   34 |         virtual ~parameter<T>() {}
      |                 ^
In file included from /home/roelof/Compiled/chess-develop/evaluate.h:13,
                 from /home/roelof/Compiled/chess-develop/evaluate.cpp:5:
/home/roelof/Compiled/chess-develop/threads.h:92:23: error: expected unqualified-id before ‘)’ token
   92 |         Threadpool<T>() { }
      |                       ^
/home/roelof/Compiled/chess-develop/threads.h:94:23: error: expected unqualified-id before ‘const’
   94 |         Threadpool<T>(const unsigned int n) :
      |                       ^~~~~
/home/roelof/Compiled/chess-develop/threads.h:94:23: error: expected ‘)’ before ‘const’
   94 |         Threadpool<T>(const unsigned int n) :
      |                      ~^~~~~
      |                       )
/home/roelof/Compiled/chess-develop/threads.h:101:9: error: template-id not allowed for destructor
  101 |         ~Threadpool<T>() { if (!stop) exit(); }
      |         ^
make[2]: *** [CMakeFiles/haVoc.dir/build.make:90: CMakeFiles/haVoc.dir/evaluate.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/haVoc.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

--- a/parameter.h
+++ b/parameter.h
@@ -23,15 +23,15 @@ protected:
 
 public:
 
-	parameter<T>(T&& in, std::string& s) : tag(s)
+	parameter(T&& in, std::string& s) : tag(s)
 	{
 		value = util::make_unique<T>(in);
 		bits = *reinterpret_cast<unsigned long*>(value.get());
 	}
-	parameter<T>(T& in, std::string& s) : tag(s) { set(in); }
-	parameter<T>(T& in) : tag("") { set(in); }
-	parameter<T>(const parameter<T>& o) { tag = o.tag;  set(*o.value); }
-	virtual ~parameter<T>() {}
+	parameter(T& in, std::string& s) : tag(s) { set(in); }
+	parameter(T& in) : tag("") { set(in); }
+	parameter(const parameter<T>& o) { tag = o.tag;  set(*o.value); }
+	virtual ~parameter() {}
 
 	parameter<T>& operator=(const parameter<T>& o) { tag = o.tag; set(*o.value); }
 	T& operator()() { return *value; }

--- a/threads.h
+++ b/threads.h
@@ -89,16 +89,16 @@ private:
 
 
 public:
-	Threadpool<T>() { }
+	Threadpool() { }
 
-	Threadpool<T>(const unsigned int n) :
+	Threadpool(const unsigned int n) :
 		busy(0), processed(0), stop(false), num_threads(n)
 	{
 		for (unsigned int i = 0; i < n; ++i)
 			workers.emplace_back(new T(std::bind(&Threadpool<T>::thread_func, this)));
 	}
 
-	~Threadpool<T>() { if (!stop) exit(); }
+	~Threadpool() { if (!stop) exit(); }
 
 	T* operator[](const int& idx) { return workers[idx]; }
 

--- a/types.h
+++ b/types.h
@@ -27,6 +27,7 @@ typedef uint8_t U8;
 
 #else
 #include <stdint.h>
+#include <cstring>
 typedef int16_t int16;
 typedef const uint64_t C64;
 typedef const uint32_t C32;

